ROS coalesce causes CS7038: Failed to emit module

**Version Used**: 
```Microsoft (R) Build Engine version 15.8.166+gd4e8d81a88 for .NET Core

.NET Core SDK (reflecting any global.json):
 Version:   2.1.402
 Commit:    3599f217f4

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.402\

Host (useful for support):
  Version: 2.1.4
  Commit:  85255dde3e
```

**Steps to Reproduce**:

1. `dotnet new classlib -f netcoreapp2.1`
2. edit Class1.cs to be:
```using System;

namespace crashtest
{
    public class Class1
    {
		private byte[] bytes = null;

		public ReadOnlySpan<byte> ByteSpan => bytes??ReadOnlySpan<byte>.Empty;
	}
}
```
3. `dotnet build`

**Expected Behavior**:
the build succeeds
**Actual Behavior**:
`CSC : error CS7038: Failed to emit module 'crashtest'.`

It can be reproduced on sharplab [here](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8kAmIA1AD4ACADAATkCMA3ALABQrAdhGHFAA4QBjONQEJoAC3ixWAb1bUFtAMy0ATNQDCAG2hQ68xXJYBIY7wSY0EeNWC54AbQC6t+z2oBeauwCuWrcxsJsbkKgBKcBAEAPLsWrgAyvzsADx28AB81ABCbkkQ7J5Z6TwA/KURUbHx+aklGQB0AKJgvDC4gcYAvqw9LEA).

--- a/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
@@ -2203,10 +2203,8 @@ moreArguments:
                 case BoundKind.NullCoalescingOperator:
                     var coalescingOp = (BoundNullCoalescingOperator)expr;
 
-                    var leftEscape = GetValEscape(coalescingOp.LeftOperand, scopeOfTheContainingExpression);
-                    var rightEscape = GetValEscape(coalescingOp.RightOperand, scopeOfTheContainingExpression);
-
-                    return Math.Max(leftEscape, rightEscape);
+                    return Math.Max(GetValEscape(coalescingOp.LeftOperand, scopeOfTheContainingExpression),
+                                    GetValEscape(coalescingOp.RightOperand, scopeOfTheContainingExpression));
 
                 case BoundKind.FieldAccess:
                     var fieldAccess = (BoundFieldAccess)expr;

--- a/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
@@ -2200,6 +2200,14 @@ moreArguments:
                     return Math.Max(consEscape,
                                     GetValEscape(conditional.Alternative, scopeOfTheContainingExpression));
 
+                case BoundKind.NullCoalescingOperator:
+                    var coalescingOp = (BoundNullCoalescingOperator)expr;
+
+                    var leftEscape = GetValEscape(coalescingOp.LeftOperand, scopeOfTheContainingExpression);
+                    var rightEscape = GetValEscape(coalescingOp.RightOperand, scopeOfTheContainingExpression);
+
+                    return Math.Max(leftEscape, rightEscape);
+
                 case BoundKind.FieldAccess:
                     var fieldAccess = (BoundFieldAccess)expr;
                     var fieldSymbol = fieldAccess.FieldSymbol;
@@ -2346,7 +2354,6 @@ moreArguments:
                 case BoundKind.AsOperator:
                 case BoundKind.AwaitExpression:
                 case BoundKind.ConditionalAccess:
-                case BoundKind.NullCoalescingOperator:
                 case BoundKind.ArrayAccess:
                     // only possible in error cases (if possible at all)
                     return scopeOfTheContainingExpression;
@@ -2502,6 +2509,11 @@ moreArguments:
                     }
 
                     return CheckValEscape(conditional.Alternative.Syntax, conditional.Alternative, escapeFrom, escapeTo, checkingReceiver: false, diagnostics: diagnostics);
+
+                case BoundKind.NullCoalescingOperator:
+                    var coalescingOp = (BoundNullCoalescingOperator)expr;
+                    return CheckValEscape(coalescingOp.LeftOperand.Syntax, coalescingOp.LeftOperand, escapeFrom, escapeTo, checkingReceiver, diagnostics) &&
+                            CheckValEscape(coalescingOp.RightOperand.Syntax, coalescingOp.RightOperand, escapeFrom, escapeTo, checkingReceiver, diagnostics);
 
                 case BoundKind.FieldAccess:
                     var fieldAccess = (BoundFieldAccess)expr;
@@ -2673,7 +2685,6 @@ moreArguments:
                 case BoundKind.AsOperator:
                 case BoundKind.AwaitExpression:
                 case BoundKind.ConditionalAccess:
-                case BoundKind.NullCoalescingOperator:
                 case BoundKind.ArrayAccess:
                     // only possible in error cases (if possible at all)
                     return false;

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
@@ -3514,5 +3514,56 @@ readonly ref struct Test
                 //         this = ref obj;
                 Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "this").WithLocation(8, 9));
         }
+
+        [Fact]
+        [WorkItem(29927, "https://github.com/dotnet/roslyn/issues/29927")]
+        public void CoalesceSpanReturn()
+        {
+            CreateCompilationWithMscorlibAndSpan(@"
+using System;
+class C
+{
+    Span<byte> M()
+    {       
+        return null ?? new Span<byte>();
+    }
+}", options: TestOptions.ReleaseDll).VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(29927, "https://github.com/dotnet/roslyn/issues/29927")]
+        public void CoalesceAssignSpanReturn()
+        {
+            CreateCompilationWithMscorlibAndSpan(@"
+using System;
+class C
+{
+    Span<byte> M()
+    {       
+        var x = null ?? new Span<byte>();
+        return x;
+    }
+}", options: TestOptions.ReleaseDll).VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(29927, "https://github.com/dotnet/roslyn/issues/29927")]
+        public void CoalesceRefSpanReturn()
+        {
+            CreateCompilationWithMscorlibAndSpan(@"
+using System;
+class C
+{
+    Span<byte> M()
+    {       
+        Span<byte> x = stackalloc byte[10];
+        return null ?? x;
+    }
+}", options: TestOptions.ReleaseDll).VerifyDiagnostics(
+                // (8,24): error CS8352: Cannot use local 'x' in this context because it may expose referenced variables outside of their declaration scope
+                //         return null ?? x;
+                Diagnostic(ErrorCode.ERR_EscapeLocal, "x").WithArguments("x").WithLocation(8, 24)
+                );
+        }
     }
 }

